Crash parsing invalid file URI

### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

We used the Fuzz engine to modify some Test(TestFromJSON) data, and then the following crash appeared. We hope to get the help of the developer to confirm whether it is a real bug.

### How to reproduce

The simplified test code is in **Example code**

### Screenshots

crash log:
```
panic: runtime error: slice bounds out of range [5:4] [recovered]
        panic: runtime error: slice bounds out of range [5:4]

goroutine 35 [running]:
testing.tRunner.func1.2({0x62e6c0, 0xc0001d43d8})
        /home/zjx/.local/go/src/testing/testing.go:1211 +0x24e
testing.tRunner.func1()
        /home/zjx/.local/go/src/testing/testing.go:1214 +0x218
panic({0x62e6c0, 0xc0001d43d8})
        /home/zjx/.local/go/src/runtime/panic.go:1038 +0x215
fyne.io/fyne/v2/storage/repository.ParseURI({0xc0001b9770, 0x4})
        /home/zjx/workspace/gowork/src/purelib/fyne/storage/repository/parse.go:63 +0x737
fyne.io/fyne/v2/storage.ParseURI(...)
        /home/zjx/workspace/gowork/src/purelib/fyne/storage/uri.go:35
fyne.io/fyne/v2/theme.uriString.resource({0xc0001b9770, 0xc00019a128})
        /home/zjx/workspace/gowork/src/purelib/fyne/theme/json.go:82 +0x1f
fyne.io/fyne/v2/theme.(*jsonTheme).Font(0xc0001b0528, {0x0, 0x0, 0x1, 0x0, 0x0})
        /home/zjx/workspace/gowork/src/purelib/fyne/theme/json.go:146 +0xbb
fyne.io/fyne/v2/theme.FuzzTestFromJSON({0xc00036f600, 0x22, 0x200})
        /home/zjx/workspace/gowork/src/purelib/fyne/theme/json_test.go_TestFromJSON_test.go:74 +0x143
```

### Example code

```go
        var1:="0000"
	var2:=`{"Fonts":{"monospace":"file"}}0000{"Fonts":{"monospace":"file"}}0000{"Fonts":{"monospace":"file"}}0000{"Fonts":{"monospace":"file"}}0000{"Fonts":{"monospace":"file"}}0000{"Fonts":{"monospace":"file"}}0000{"Fonts":{"monospace":"file"}}0000{"F`
	repository.Register(var1, intRepo.NewFileRepository())
	th, err := FromJSON(var2)
	assert.Nil(t, err)
	th.Font(fyne.TextStyle{Monospace: true}).Name()
```

### Fyne version

638ae24e4bebfc0b6dbaabf8db5807eeabfafc61

### Go compiler version

1.17.8

### Operating system

Linux

### Operating system version

Linux r920 4.15.0-159-generic #167-Ubuntu

### Additional Information

_No response_

--- a/storage/repository/parse.go
+++ b/storage/repository/parse.go
@@ -44,20 +44,12 @@ func NewFileURI(path string) fyne.URI {
 // Since: 2.0
 func ParseURI(s string) (fyne.URI, error) {
 	// Extract the scheme.
-	scheme := ""
-	validScheme := false
-	for i := 0; i < len(s); i++ {
-		if s[i] == ':' {
-			validScheme = true
-			break
-		}
-		scheme += string(s[i])
-	}
-	scheme = strings.ToLower(scheme)
-
-	if !validScheme {
+	colonIndex := strings.Index(s, ":")
+	if colonIndex == -1 {
 		return nil, errors.New("invalid URI, scheme must be present")
 	}
+
+	scheme := strings.ToLower(s[:colonIndex])
 
 	if scheme == "file" {
 		// Does this really deserve to be special? In principle, the

--- a/storage/repository/parse.go
+++ b/storage/repository/parse.go
@@ -44,8 +44,8 @@ func NewFileURI(path string) fyne.URI {
 // Since: 2.0
 func ParseURI(s string) (fyne.URI, error) {
 	// Extract the scheme.
-	colonIndex := strings.Index(s, ":")
-	if colonIndex == -1 {
+	colonIndex := strings.IndexByte(s, ':')
+	if colonIndex <= 0 {
 		return nil, errors.New("invalid URI, scheme must be present")
 	}
 

--- a/storage/repository/parse.go
+++ b/storage/repository/parse.go
@@ -1,6 +1,7 @@
 package repository
 
 import (
+	"errors"
 	"path/filepath"
 	"runtime"
 	"strings"
@@ -44,13 +45,19 @@ func NewFileURI(path string) fyne.URI {
 func ParseURI(s string) (fyne.URI, error) {
 	// Extract the scheme.
 	scheme := ""
+	validScheme := false
 	for i := 0; i < len(s); i++ {
 		if s[i] == ':' {
+			validScheme = true
 			break
 		}
 		scheme += string(s[i])
 	}
 	scheme = strings.ToLower(scheme)
+
+	if !validScheme {
+		return nil, errors.New("invalid URI, scheme must be present")
+	}
 
 	if scheme == "file" {
 		// Does this really deserve to be special? In principle, the

--- a/storage/repository/parse_test.go
+++ b/storage/repository/parse_test.go
@@ -24,3 +24,17 @@ func TestParseURI(t *testing.T) {
 	assert.Nil(t, err)
 	assert.Equal(t, "file://C:/tmp/foo.txt", uri.String())
 }
+
+func TestParseInvalidURI(t *testing.T) {
+	uri, err := ParseURI("/tmp/foo.txt")
+	assert.NotNil(t, err)
+	assert.Nil(t, uri)
+
+	uri, err = ParseURI("file")
+	assert.NotNil(t, err)
+	assert.Nil(t, uri)
+
+	uri, err = ParseURI(":foo")
+	assert.NotNil(t, err)
+	assert.Nil(t, uri)
+}

